iceoryx_posh/include/iceoryx_posh/popo/sample.hpp

```diff
@@ -81,69 +134,51 @@ class Sample
     operator bool() const;
 
     ///
-    /// @brief allocation Access to the memory chunk loaned to the sample.
-    /// @return
+    /// @brief allocation Access to the encapsulated type loaned to the sample.
+    /// @return a pointer to the encapsulated type.
+    /// @details Only available for non-const type T.
     ///
+    template <typename S = T, typename = std::enable_if_t<std::is_same<S, T>::value && !std::is_const<S>::value, S>>
     T* get() noexcept;
 
     ///
-    /// @brief allocation Read-only access to the memory chunk loaned to the sample.
-    /// @return
+    /// @brief allocation Read-only access to the encapsulated type loaned to the sample.
+    /// @return a const pointer to the encapsulated type.
     ///
     const T* get() const noexcept;
 
     ///
     /// @brief header Retrieve the header of the underlying memory chunk loaned to the sample.
     /// @return The ChunkHeader of the underlying memory chunk.
+    /// @details Only available for non-const type T.
     ///
+    template <typename S = T, typename = std::enable_if_t<std::is_same<S, T>::value && !std::is_const<S>::value, S>>
     mepoo::ChunkHeader* getHeader() noexcept;
 
+    ///
+    /// @brief header Retrieve the header of the underlying memory chunk loaned to the sample.
+    /// @return The const ChunkHeader of the underlying memory chunk.
+    ///
+    const mepoo::ChunkHeader* getHeader() const noexcept;
+
     ///
     /// @brief publish Publish the sample via the publisher from which it was loaned and automatically
     /// release ownership to it.
+    /// @details Only available for non-const type T.
     ///
+    template <typename S = T, typename = std::enable_if_t<std::is_same<S, T>::value && !std::is_const<S>::value, S>>
     void publish() noexcept;
 
     ///
     /// @brief release Manually release ownership of the loaned memory chunk.
-    /// @details This prevents the sample from automatically releasing ownership on destruction.
+    /// @details This prevents the sample from automatically releasing ownership on destruction and is only available
+    /// for non-const type T.
     ///
+    template <typename S = T, typename = std::enable_if_t<std::is_same<S, T>::value && !std::is_const<S>::value, S>>
```

Comment:
Is `static_assert` possible?

Reply:
Previously this was only available for the sample when loaned from a publisher and probably also a remnant of the previous refactoring with the `void` sample. I would suggest to get rid of this method, since it makes the `Sample` explicitly null-able. I would suggest to use a `cxx::optional<Sample<T>>` if it is necessary to store a sample for later use and being able to release it.

@ithier what do you think?

Reply:
Yes, I also noticed this right a few minutes ago. It probably should not be public since it can lead to chunk leaks.sample
I will revert the change to not `reset` the pointer but `release` it. This Method should probably be private and the `Publisher` a friend of `Sample`.